comp110-worksheet-6/OxoBoard.cs

```diff
@@ -10,38 +10,136 @@ public enum Mark { None, O, X };
+                for (int j = 0; j < width; j++)
+                {
+                    sublist.Add(Mark.None);
+                }
+                board.Add(sublist);
+            }
 		}
 
 		// Return the contents of the specified square.
 		public Mark GetSquare(int x, int y)
 		{
-			throw new NotImplementedException("TODO: implement this function and then remove this exception");
+            try
+            {
+                return board[x][y];
+            }
+            catch (System.ArgumentOutOfRangeException)
+            {
+                return Mark.X;
+            }
 		}
 
 		// If the specified square is currently empty, fill it with mark and return true.
 		// If the square is not empty, leave it as-is and return False.
 		public bool SetSquare(int x, int y, Mark mark)
 		{
-			throw new NotImplementedException("TODO: implement this function and then remove this exception");
+            if (GetSquare(x, y) == Mark.None)
+            {
+                board[x][y] = mark;
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+            
 		}
 
 		// If there are still empty squares on the board, return false.
 		// If there are no empty squares, return true.
 		public bool IsBoardFull()
 		{
-			throw new NotImplementedException("TODO: implement this function and then remove this exception");
+            bool hasSpace = true;
+            foreach (List<Mark> row in board)
+            {
+                foreach (Mark mark in row)
+                {
+                    if (mark == Mark.None)
+                    {
+                        hasSpace = false;
+                    }
+                }
+            }
+            return hasSpace;
 		}
 
+        private Mark checkLine(List<Mark> line)
+        {
+            Mark winner = line[0];
+            Mark firstMark = line[0];
+            foreach (Mark mark in line)
+            {
+                if ((firstMark == Mark.None) || (mark != firstMark))
+                {
+                    winner = Mark.None;
+                }
+            }
+            return winner;
+        }
+
 		// If a player has three in a row, return Mark.O or Mark.X depending on which player.
 		// Otherwise, return Mark.None.
 		public Mark GetWinner()
 		{
-			throw new NotImplementedException("TODO: implement this function and then remove this exception");
+            Mark winner = Mark.None;
+
+            // Check the Columns of the board.
+            foreach (List<Mark> line in board)
+            {
+                if (winner == Mark.None)
+                {
+                    winner = checkLine(line);
+                }
+            }
+
+            
+            // Check the Rows of the board.
+            for (int i = 0; i < board.Count; i++)
+            {
+                if (winner == Mark.None)
+                {
+                    List<Mark> row = new List<Mark>();
```

Comment:
Slight inefficiency here -- you're allocating quite a few lists. It would be more efficient to reuse them, or to make use of something like iterator functions to avoid allocating entirely. (Doesn't really matter in this exercise but worth bearing in mind for more performance-critical games work.)